mem/buffer_pool.go

```diff
@@ -0,0 +1,184 @@
+	// Get returns a buffer with specified length from the pool.
+	Get(length int) []byte
+
+	// Put returns a buffer to the pool.
+	Put([]byte)
+}
+
+var defaultBufferPoolSizes = []int{
+	256,
+	4 << 10,  // 4KB (go page size)
+	16 << 10, // 16KB (max HTTP/2 frame size used by gRPC)
+	32 << 10, // 32KB (default buffer size for io.Copy)
+	// TODO: Report the buffer sizes requested with Get to tune this properly.
+	1 << 20, // 1MB
+}
+
+var defaultBufferPool = func() *atomic.Pointer[BufferPool] {
+	pool := NewBufferPool(defaultBufferPoolSizes...)
+	ptr := new(atomic.Pointer[BufferPool])
+	ptr.Store(&pool)
+	return ptr
+}()
+
+// DefaultBufferPool returns the current default buffer pool. It is a BufferPool
+// created with NewBufferPool that uses a set of default sizes optimized for
+// expected workflows.
+func DefaultBufferPool() BufferPool {
+	return *defaultBufferPool.Load()
+}
+
+// SetDefaultBufferPoolForTesting updates the default buffer pool, for testing
+// purposes.
+func SetDefaultBufferPoolForTesting(pool BufferPool) {
+	defaultBufferPool.Store(&pool)
+}
+
+// NewBufferPool returns a BufferPool implementation that uses multiple
+// underlying pools of the given pool sizes. When a buffer is requested from the
+// returned pool, it will have a
+func NewBufferPool(poolSizes ...int) BufferPool {
+	sort.Ints(poolSizes)
+	pools := make([]*sizedBufferPool, len(poolSizes))
+	for i, s := range poolSizes {
+		pools[i] = newBufferPool(s)
+	}
+	return &tieredBufferPool{
+		sizedPools: pools,
+	}
+}
+
+// tieredBufferPool implements the BufferPool interface with multiple tiers of
+// buffer pools for different sizes of buffers.
+type tieredBufferPool struct {
+	sizedPools   []*sizedBufferPool
+	fallbackPool simpleBufferPool
+}
+
+func (p *tieredBufferPool) Get(size int) []byte {
+	return p.getPool(size).Get(size)
+}
+
+func (p *tieredBufferPool) Put(buf []byte) {
+	p.getPool(cap(buf)).Put(buf)
+}
+
+func (p *tieredBufferPool) getPool(size int) BufferPool {
+	poolIdx := sort.Search(len(p.sizedPools), func(i int) bool {
+		return p.sizedPools[i].defaultSize >= size
+	})
+
+	if poolIdx == len(p.sizedPools) {
+		return &p.fallbackPool
+	}
+
+	return p.sizedPools[poolIdx]
+}
+
+// sizedBufferPool is a BufferPool implementation that is optimized for specific
+// buffer sizes. For example, HTTP/2 frames within grpc are always 16kb and a
+// sizedBufferPool can be configured to only return buffers with a capacity of
+// 16kb. Note that however it does not support returning larger buffers and in
+// fact panics if such a buffer is requested.
+type sizedBufferPool struct {
+	pool        sync.Pool
+	defaultSize int
+}
+
+func (p *sizedBufferPool) Get(size int) []byte {
+	bs := *p.pool.Get().(*[]byte)
+	return bs[:size]
+}
+
+func (p *sizedBufferPool) Put(buf []byte) {
+	if cap(buf) < p.defaultSize {
+		// Ignore buffers that are too small to fit in the pool. Otherwise, when Get is
+		// called it will panic as it tries to index outside the bounds of the buffer.
```

Comment:
Would it instead be better for `Get` to handle the case where the buffer returned from `p.pool.Get()` is of size less than `size`, by returning `nil`. Because IIUC, if `Get` returns `nil`, then `sync.Pool` would call `New()`. That seems to me like a better behavior than panicking.

Reply:
Thanks. I do the see docs being update on the `sizedBufferPool` type. That should be good.